integration-tests/src/tests/client/features/congestion_control.rs

```diff
@@ -0,0 +1,228 @@
+use near_chain_configs::Genesis;
+use near_client::test_utils::TestEnv;
+use near_client::ProcessTxResponse;
+use near_crypto::{InMemorySigner, KeyType, PublicKey};
+use near_primitives::account::id::AccountId;
+use near_primitives::errors::{ActionErrorKind, FunctionCallError, TxExecutionError};
+use near_primitives::shard_layout::ShardLayout;
+use near_primitives::transaction::SignedTransaction;
+use near_primitives::version::{ProtocolFeature, PROTOCOL_VERSION};
+use near_primitives::views::FinalExecutionStatus;
+use nearcore::test_utils::TestEnvNightshadeSetupExt;
+
+fn setup_runtime(sender_id: AccountId) -> TestEnv {
+    let epoch_length = 10;
+    let mut genesis = Genesis::test_sharded_new_version(vec![sender_id], 1, vec![1, 1, 1, 1]);
+    genesis.config.epoch_length = epoch_length;
+    genesis.config.protocol_version = ProtocolFeature::CongestionControl.protocol_version() - 1;
+    // Chain must be sharded to test cross-shard congestion control.
+    genesis.config.shard_layout = ShardLayout::v1_test();
+    TestEnv::builder(&genesis.config).nightshade_runtimes(&genesis).build()
+}
+
+/// Simplest possible upgrade to new protocol with congestion control enabled,
+/// no traffic at all.
+#[test]
+fn test_protocol_upgrade() {
+    // The following only makes sense to test is the feature is enabled in the current build.
+    if !ProtocolFeature::CongestionControl.enabled(PROTOCOL_VERSION) {
+        return;
+    }
+
+    // The immediate protocol upgrade needs to be set for this test to pass in
+    // the release branch where the protocol upgrade date is set.
+    std::env::set_var("NEAR_TESTS_IMMEDIATE_PROTOCOL_UPGRADE", "1");
+
+    let mut env = setup_runtime("test0".parse().unwrap());
+
+    // Produce a few blocks to get out of initial state.
+    let tip = env.clients[0].chain.head().unwrap();
+    for i in 1..4 {
+        env.produce_block(0, tip.height + i);
+    }
+
+    // Ensure we are still in the old version and no congestion info is shared.
+    check_old_protocol(&env);
+
+    // The following only makes sense to test is the feature is enabled in the current build.
+    if !ProtocolFeature::CongestionControl.enabled(PROTOCOL_VERSION) {
+        return;
+    }
+    env.upgrade_protocol_to_latest_version();
+
+    // check we are in the new version
+    let block = env.clients[0].chain.get_head_block().unwrap();
```

Comment:
micro nit - Since you read the block here I was expecting that it'll be used for the check in the next line but actually you do that with env. Can you move this line two lines below?

Reply:
true, too much frantic moving around of code :)